projects/laji/src/app/+theme/kerttu/kerttu-result/kerttu-letter-result-table/kerttu-letter-result-table.component.ts

```diff
@@ -0,0 +1,104 @@
+import {Component, OnInit, ChangeDetectionStrategy, Input, OnDestroy, ChangeDetectorRef} from '@angular/core';
+import {IUserLetterTaxonStatistics} from '../../models';
+import {DatatableColumn} from '../../../../shared-modules/datatable/model/datatable-column';
+import { TranslateService } from '@ngx-translate/core';
+import { TriplestoreLabelService } from '../../../../shared/service/triplestore-label.service';
+import {Subscription} from 'rxjs';
+
+interface ILetterResultTableRow extends IUserLetterTaxonStatistics {
+  species?: string;
+}
+
+@Component({
+  selector: 'laji-kerttu-letter-result-table',
+  templateUrl: './kerttu-letter-result-table.component.html',
+  styleUrls: ['./kerttu-letter-result-table.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class KerttuLetterResultTableComponent implements OnInit, OnDestroy {
+  columns: DatatableColumn[] = [
+    {
+      name: 'species',
+      label: 'theme.kerttu.result.species',
+      width: 100,
+      summaryFunc: () => this.translate.instant('theme.total')
+    },
+    {
+      name: 'userAnnotationCount',
+      label: 'theme.kerttu.result.userAnnotationCount',
+      width: 70
+    },
+    {
+      name: 'commonAnnotationCount',
+      label: 'theme.kerttu.result.commonAnnotationCount',
+      width: 70
+    },
+    {
+      name: 'similarity',
+      label: 'theme.kerttu.result.similarity',
+      cellTemplate: 'percentage',
+      width: 70,
+      summaryFunc: this.meanPercent
+    },
+    {
+      name: 'identifiability',
+      label: 'theme.kerttu.result.identifiability',
+      cellTemplate: 'percentage',
+      width: 70,
+      summaryFunc: this.meanPercent
+    }
+  ];
+  _data: ILetterResultTableRow[] = [];
+
+  loading = false;
+
+  @Input() set data(data: IUserLetterTaxonStatistics[]) {
+    this._data = [...data];
+    this.updateSpecies();
+  }
+
+  private nameSub: Subscription;
+
+  constructor(
+    private translate: TranslateService,
+    private triplestoreLabelService: TriplestoreLabelService,
+    private cd: ChangeDetectorRef
+  ) { }
+
+  ngOnInit(): void {
+  }
+
+  ngOnDestroy() {
+    if (this.nameSub) {
+      this.nameSub.unsubscribe();
+    }
+  }
+
+  updateSpecies() {
+    if (this.nameSub) {
+      this.nameSub.unsubscribe();
+    }
+    const ids = this._data.map(d => d.taxonId);
+    if (ids.length < 1) {
+      return;
+    }
+
+    this.loading = true;
+    this.nameSub = this.triplestoreLabelService.getAll(ids, this.translate.currentLang).subscribe(res => {
+      this._data.forEach(d => { d.species = res[d.taxonId]; });
+      this._data = [...this._data];
+      this.loading = false;
+      this.cd.markForCheck();
+    });
+  }
```

Comment:
Maybe using a cell template for the species label would be more elegant? (no need for creating pipe, subscribing/unsubscribing, resetting _data & touching the change detector...). I couldn't find an existing pipe for this by taking a quick glance, but an example solution is the winter bird species result service (`projects/laji/src/app/+project-form/results/wbc-result/wbc-species/wbc-species-list/wbc-species-list.component.ts`)

Reply:
There already exists a label pipe but the problem is that the automatic sorting doesn't work then. Some kind of custom sorting would need to be implemented (like in the species list `projects/laji/src/app/+taxonomy/species/species-list/species-list.component.ts`). This is probably quite a common problem so it should be on the laji-datatable component. But it sounds like a quite big task since it would possibly affect other tables so maybe it should be its own task?

Reply:
Ok!